### PR TITLE
Allow notifying on every build, or every failure + status change.

### PR DESCRIPTION
**Pull request by alboyadjian**

### CODE REVIEW

**Comment by alboyadjian** [comment]
@davexunit @estebanz01 Would appreciate feedback on the following:
1. Is the correct default for DEFAULT_NOTIFY_STRATEGY `status_change`, or should I make it `every_failure_and_status_change` as that seems to be the consensus on what's desired based on the slack poll.
2. Is `every_failure_and_status_change` a good enough name? I also considered: `status_changes_and_failures` `fail_or_status_change` `status_change_or_fail`
3. In order to have a default strategy and then have certain branches for which a different strategy is used, I chose to have an ENV var named after the strategy that takes an array of branch names for which that override applies. I couldn't think of a better way that'd work using env vars and command line. I considered using a pattern like `--strategy-for-master=every_build` where you could specify branch-by-branch, but translating between command-line args and env vars seemed cumbersome. I also considered some kind of key-value pair structure as the argument, e.g.
```
--override-strategies=master:every_build,release:every_build
```
4. If the approach I took here works, any suggestions for the name of the override arg?
`'--notify-every-build-branches=BRANCHES',` makes sense if you know what it is, but I don't think it's intuitive. Maybe "branches" isn't necessary.
5. Should I go ahead and add support for the override args/env vars for the other two strategies?

**Comment by estebanz01** [comment]
1. I think we should go with the consensus.
2. I think it makes sense to name it: `fail_or_status_change`. It gives the user a basic idea of what's going on when you do `DEFAULT_NOTIFY_STRATEGY=fail_or_status_change` while keeping it relatively short.
3. I like that key-pair approach. I shouldn't be a lot of branches with override strategies, I assume.
So, 4 and 5. does not apply for my feedback. 🙂

File: bin/update-build-status [hunk 1]
```diff
@@ -63,16 +63,31 @@ OptionParser.new do |opts|
...
   opts.on(
-    '--slack-admin-usernames=USERS',
-    'comma-separated list of slack users to be notified if build status ' \
-    'notifications fail to send'
-  ) { |usernames| command_line_opts[:slack_admin_users] = usernames }
+    '--default-notify-strategy=STRATEGY',
+    'when to send notifications, in the absence of a branch-specific ' \
+    'override -- valid options are: status_change (the default), ' \
+    'every_build, fail_or_status_change'
+  ) { |default_strategy| command_line_opts[:default_strategy] = default_strategy }
```
Comment (davexunit) [inline bin/update-build-status]: Not a big deal but any reason why you chose underscores and not dashes for the strategy names?

Comment (alboyadjian) [reply bin/update-build-status]: I thought the underscores would make it easier to visualize the difference between the command-line flag and the values.
```
--default-notify-strategy=every_build
--default-notify-strategy=every-build
```
If you think that's unnecessary, I can change it.

File: bin/update-build-status [hunk 2]
```diff
@@ -63,16 +63,30 @@ OptionParser.new do |opts|
...
     '--dynamo-table=TABLE',
     'table for storing build statuses'
   ) { |table| command_line_opts[:dynamo_table] = table }
 
+  opts.on(
+    '--notify-every-build-branches=BRANCHES',
+    'allows branch-specific override for default notify strategy; ' \
+    'comma-separated list of branches that will have build notifications ' \
+    'sent for every build, regardless of previous build status'
+  ) { |branches| command_line_opts[:notify_every_build_branches] = branches }
```
[the remark this answers is not in the document]
Comment (alboyadjian) [reply bin/update-build-status]: No reason, will fix.

**Pull request merged**